### PR TITLE
Import Licence

### PR DESCRIPTION
**Pull request by jonathangoulding**
We need to replace the import service logic to import a licence from NALD.

We will leave the import service running to add the NALD data in the import tables, but we will slowly replace the import-licence functionality and move this into the system repo.

https://eaflood.atlassian.net/browse/WATER-4575

This change will add an endpoint to receive the licence id. With an associated controller and services to get the data from the import tables, transform the data into the format needed to store in the relevant licence tables and persist the licence.

This should create and/or update the licence as per the current import-licence process.

### CODE REVIEW

File: app/controllers/import.controller.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Controller for /import
+ * @module ImportController
+ */
+async function licence (request, h) {
+  const { licenceRef } = request.payload
+
+  await ImportLicenceService.go(licenceRef)
+
+  return h.response().code(204)
```
Comment (jonathangoulding) [inline app/controllers/import.controller.js:14]: Try catch here could allow a different error code.

File: app/routes/import.routes.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+const routes = [
+  {
+    method: 'POST',
+    path: '/import/licence',
+    handler: ImportController.licence,
+    options: {
+      app: {
+        plainOutput: true
+      },
+      auth: false
```
Comment (jonathangoulding) [inline app/routes/import.routes.js:14]: auth needed ?

File: app/services/import/fetch-licence-from-import.service.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+async function go (licenceRef) {
+  const licence = await _getLicenceByRef(licenceRef)
+
+  return {
+    licence
+  }
+}
+
+async function _getLicenceByRef (licenceRef) {
+  // TODO: granular select  ?
```
Comment (jonathangoulding) [inline app/services/import/fetch-licence-from-import.service.js:18]: Map to the input and mapper requirements.

File: app/services/import/helpers/dates.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  const [day, month, year] = date.split('/')
+
+  const isoDateString = `${year}-${month}-${day}`
+
+  if (isValidDate(isoDateString)) {
+    return isoDateString
+  } else {
+    //  TODO: what do we want to do ? allow the top catch to log the error
```
Comment (jonathangoulding) [inline app/services/import/helpers/dates.js:16]: Should it fail

File: app/services/import/licence.service.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    console.log('Mapped licence data', mappedLicenceData)
+    // validate
+
+    // persist
+    const savedLicence = await PersistLicenceService.go(mappedLicenceData)
+
+    console.log('Saved licence', savedLicence)
+  } catch (e) {
+    //  Should we do this in the controller and pass the message back to import service ?
+    console.error('Licence', licenceRef, 'failed to import: ', e)
```
Comment (jonathangoulding) [inline app/services/import/licence.service.js:36]: Proper logger

File: app/services/import/mappers/import-service/licence.mapper.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  return { historicalAreaCode, regionalChargeArea, standardUnitChargeCode, localEnvironmentAgencyPlanCode }
+}
+
+const getMinDate = (arr) => {
+  const sorted = getSortedDates(arr)
+
+  return sorted.length === 0 ? null : sorted[0]
+}
+
+const getSortedDates = (arr) => {
```
Comment (jonathangoulding) [inline app/services/import/mappers/import-service/licence.mapper.js:88]: lift into dates - generic sort

File: app/services/import/mappers/import-service/licence.mapper.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  const historicalAreaCode = licenceData.AREP_AREA_CODE
+  const regionPrefix = licenceData.AREP_EIUC_CODE.substr(0, 2)
+  const regionalChargeArea = regions[regionPrefix]
+  const standardUnitChargeCode = licenceData.AREP_SUC_CODE
+  const localEnvironmentAgencyPlanCode = licenceData.AREP_LEAP_CODE
+
+  return { historicalAreaCode, regionalChargeArea, standardUnitChargeCode, localEnvironmentAgencyPlanCode }
+}
+
+const getMinDate = (arr) => {
```
Comment (jonathangoulding) [inline app/services/import/mappers/import-service/licence.mapper.js:82]: lift into dates

File: app/services/import/persist-licence.service.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+ * @module PersistLicenceService
+ */
+
+const LicenceModel = require('../../models/licence.model.js')
+const RegionModel = require('../../models/region.model.js')
+
+async function go (licence) {
+  const {
+    licence: {
+      regionCode,
```
Comment (jonathangoulding) [inline app/services/import/persist-licence.service.js:14]: put this straight in if no diff with the mapper. 

region might be issue and other things on the licence might not work

File: app/services/import/mappers/import-service/licence.mapper.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * @return {String} YYYY-MM-DD
+ */
+const mapStartDate = (licence, licenceVersions) => {
+  if (licence.ORIG_EFF_DATE !== 'null') {
+    return formatNaldToISO(licence.ORIG_EFF_DATE)
+  }
+
+  // TODO: assume a licence versions must exists if no start date ORIG_EFF_DATE ?
+  // Not draft is handle in the database query ? should we keep here for resps
+  return licenceVersions
```
Comment (jonathangoulding) [inline app/services/import/mappers/import-service/licence.mapper.js:74]: This used to remove status DRAFT but this is done in the SQL query should we still keep this ?

File: app/services/import/licence-validator.service.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+/**
+ * Validates a licence is in the correct shape and format to perist in the databasse
+ *
+ * @param {ValidLicenceType} licence - The licence to validator
+ * @returns {} a licence is the licence is valid, other wise throw an error
+ */
+function go (licence) {
+  return licence
```
Comment (jonathangoulding) [inline app/services/import/licence-validator.service.js:15]: Implement validator

File: app/services/import/persist-licence.service.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Persists the licence
+ * @module PersistLicenceService
+ */
+
+const LicenceModel = require('../../models/licence.model.js')
+const RegionModel = require('../../models/region.model.js')
+
+async function go (licence) {
```
Comment (jonathangoulding) [inline app/services/import/persist-licence.service.js:11]: test persist - don't for get the insert / or create

File: app/lib/dates.lib.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  return !isNaN(date.getTime())
+}
+
+/**
+ * Checks a string matches the ISO 8601 date format
+ *
+ * @param {dateString} dateString - The date in the iso format 2001-01-01
+ * @returns {boolean}
+ */
+function isISODateFormat (dateString) {
```
Comment (jonathangoulding) [inline app/lib/dates.lib.js:64]: I think this should throw an error.

I think we want this functionally to highlight any date issues. 

And we place it wherever a date string should be in the correct format. (mainly in this dates lib)

**Review by robertparkinson — APPROVED** [review]

**Pull request merged**